projects/observability/src/shared/components/topology/d3/layouts/graph-layout.ts

```diff
@@ -0,0 +1,76 @@
+import { RenderableTopology, RenderableTopologyNode, TopologyCoordinates, TopologyEdge, TopologyNode } from "../../topology";
+
+export class GraphLayout {
+  public layout(topology: RenderableTopology<TopologyNode, TopologyEdge>): void {
+
+    const levelToNodesMap = new Map<number, RenderableTopologyNode[]>();
+    const nodeToLevelMap = new Map<RenderableTopologyNode, number>();
+    const nodeCoOrdinatesMap: Map<string, TopologyCoordinates> = new Map<string, TopologyCoordinates>();
+
+    this.findRootNodes(topology, levelToNodesMap, nodeToLevelMap);
+    this.fillNodeAndLevelMaps(levelToNodesMap, nodeToLevelMap);
+    this.fillNodeCoOrdinatesMap(nodeCoOrdinatesMap, levelToNodesMap);
+    this.setNodeCoOrdinates(topology, nodeCoOrdinatesMap);
+  }
+
+  private findRootNodes(topology: RenderableTopology<TopologyNode, TopologyEdge>, levelToNodesMap: Map<number, RenderableTopologyNode[]>, nodeToLevelMap: Map<RenderableTopologyNode, number>) {
+    topology.nodes.forEach(node => {
+      if (node.incoming.length === 0) {
+        nodeToLevelMap.set(node, 0);
+        levelToNodesMap.has(0) ? levelToNodesMap.get(0)?.push(node) : levelToNodesMap.set(0, [node])
+      }
+    });
+  }
+
+  private fillNodeAndLevelMaps(levelToNodesMap: Map<number, RenderableTopologyNode[]>, nodeToLevelMap: Map<RenderableTopologyNode, number>) {
+
+    const rootNodes: RenderableTopologyNode[] = levelToNodesMap.get(0) ?? [];
+    const visited = new Set<string>();
```

Comment:
we're basing it on the name. Are we sure that name is a unique string every time? I hope so but we can't be sure. We can use nodes for this and as these are the objects stored by reference. So if we do that, I think it should work as expected.

Reply:
You can also do this, every time a new node is created, generate a unique id for that node and use it. There's a `uniqueId()` in the loadash

Reply:
Oh yeah right, I converted it that way to use name in cytoscape as it accepts only strings and not objects for nodes and ended up using the names here as well, while here we can very well work with nodes as well. 
Thanks for that!